src/main/scala/io/iohk/ethereum/domain/BlockchainReader.scala

```diff
@@ -113,6 +110,38 @@ class BlockchainReader(
   def genesisBlock: Block =
     getBlockByNumber(0).get
 
+  /** Returns a block inside this branch based on its number */
+  def getBlockByNumber(branch: Branch, number: BigInt): Option[Block] = branch match {
+    case BestBranch(_, tipBlockNumber) =>
+      if (tipBlockNumber >= number && number >= 0) {
+        for {
+          hash <- getHashByBlockNumber(number)
+          block <- getBlockByHash(hash)
+        } yield block
+      } else None
+    case EmptyBranch => None
+  }
+
+  /** Returns a block hash for the block at the given height if any */
+  def getHashByBlockNumber(branch: Branch, number: BigInt): Option[ByteString] = branch match {
```

Comment:
what do you think of using a type alias for ByteString when we are talking about a hash?

Reply:
I would prefer a value class or a tagged type, because the type alias would just be "cosmetic". But it would probably be out of scope to introduce the new type in this PR.